### PR TITLE
closes GH-18, adds api to get nearest breweries

### PR DESCRIPTION
**Pull request by noebrito**
Adds support for querying for nearest breweries by a point radius measured in meters. Using Postgis to perform the geometric computation.

### CODE REVIEW

File: src/main/java/org/openbrewerydb/AppServer.java [hunk 1]
```diff
@@ -35,6 +36,9 @@ public void run(final int port) {
         path(":id", () -> {
           get(this.breweryApi::getBrewery);
         });
+        path("search-nearest", () -> {
+          post(this.breweryApi::getNearestBreweries);
+        });
```
Comment (noebrito) [inline src/main/java/org/openbrewerydb/AppServer.java]: I honestly don't have any experience with rest endpoints or the conventions here. But I want to be able to provide a request body so I assume it's ok to use a post?

Comment (Johnnyk737) [reply src/main/java/org/openbrewerydb/AppServer.java]: POST is what I've used in the past so I'm ok with it.

Comment (chrisjm) [reply src/main/java/org/openbrewerydb/AppServer.java]: I think this should be GET. 

https://blog.teamtreehouse.com/the-definitive-guide-to-get-vs-post

- Rule #1: Use GET for safe actions and POST for unsafe actions.
- Rule #2: Use POST when dealing with sensitive data.
- Rule #3: Use POST when dealing with long requests.

I don't think any of those apply, but let me know if I'm wrong.

**Review by chrisjm — CHANGES_REQUESTED** [review]
Requesting POST to be changed to GET. Otherwise, looks good! I really need to learn PostGIS queries.

File: src/main/java/org/openbrewerydb/AppServer.java (hunk 1 above)
Comment (noebrito) [reply src/main/java/org/openbrewerydb/AppServer.java]: How do I pass in my object and radius fields as a request?

Comment (noebrito) [reply src/main/java/org/openbrewerydb/AppServer.java]: Libraries don't allow passing in a request body for GET requests. Passing in as query parameters is I guess possible but not as user friendly IMO (in terms of input validation).

Comment (noebrito) [reply src/main/java/org/openbrewerydb/AppServer.java]: @chrisjm this is a good read and is the same problem. Haha. 

https://dropbox.tech/developers/limitations-of-the-get-method-in-http

Comment (chrisjm) [reply src/main/java/org/openbrewerydb/AppServer.java]: I'm not really following the "not as user friendly" argument. The current API works with query parameters. Can you expand on that?

Comment (chrisjm) [reply src/main/java/org/openbrewerydb/AppServer.java]: Also, not sure I follow the relevance of the Dropbox article.

> The Dropbox API tries to use GET for calls that don't modify server state, but unfortunately this isn't always possible. GET requests don't have a request body, so all parameters must appear in the URL or in a header. While the HTTP standard doesn't define a limit for how long URLs or headers can be, mostHTTP clients and servers have a practical limit somewhere between 2 kB and 8 kB.

As far as I can tell we're only sending lat, long, and a number. What am I missing?

Comment (noebrito) [reply src/main/java/org/openbrewerydb/AppServer.java]: I'm fine adding it as query params, but it does require the server to implement custom input validation logic whereas as a request body you effectively get your input validation for free via JSON deserialization.

Just provided the Dropbox link as an example from a top tech company where conventions can be broken.

And to reiterate, I don't use rest APIs so I myself am learning best practices.

Comment (chrisjm) [reply src/main/java/org/openbrewerydb/AppServer.java]: I think custom input validation is great. Rarely in the real world do you get anything for free :)

File: src/test/java/org/openbrewerydb/api/BreweryApiTest.java [hunk 2]
```diff
@@ -79,4 +84,27 @@ public void testGetBrewery() {
...
+    request = new GetNearestBreweriesRequest(new Location(1.0, 1.0), 1);
+    response = Unirest.post(String.format("%s/search-nearest", this.apiUrl))
+        .body(JavalinJson.toJson(request))
+        .asString();
+
+    assertThat(response.getStatus()).isEqualTo(200);
+    breweries = JavalinJson.fromJson(response.getBody(), GetNearestBreweriesResponse.class).breweries();
+    assertThat(breweries).isEmpty();
+  }
+
```
Comment (Johnnyk737) [inline src/test/java/org/openbrewerydb/api/BreweryApiTest.java:109]: Do you think we could have some tests where we are validating when 1/3 or 2/3 parameters are passed in and throw an appropriate error? Usually, that should be a 400

File: src/main/java/org/openbrewerydb/AppServer.java (hunk 1 above)
Comment (Johnnyk737) [reply src/main/java/org/openbrewerydb/AppServer.java]: I threw in a comment for the tests, but if the input fails validation, we should throw a 400.

**Review by chrisjm — APPROVED** [review]
Awesome! I dig the separate Exceptions class. Ship it!

**Pull request merged**